### ISSUE TITLE
Blazor server-side sample?

### ISSUE THREAD

**Issue by win32nipuh** [Finbuckle/Finbuckle.MultiTenant#250]
It will be very useful to have an example how to use  Finbuckle.MultiTenant in a Blazor Web Server. Do you have a plan to create one?

**Comment by AndrewTriesToCode** [comment]
hi @win32nipuh 

I would like to, but I need to get up-to-date with Blazor and so it probably won't be in the immediate future. There are some decent discussions about it in the closed issues if you haven't found them yet.

**Comment by win32nipuh** [comment]
Thanx. I see ok. 
Andrew could you please take a look at my another question? 
I need to pass multiple tenants parameters  like Connection string to database, S3 connection parameters etc.
How I can do it?

**Comment by ZedZipDev** [comment]
Hi Andrew,
I have created full functional Blazor Server app DataIsolationBlazorSample
https://github.com/Oleg26Dev/Finbuckle.MultiTenant/tree/release-6.0.0-dev/samples/ASP.NET%20Core%203/DataIsolationBlazorSample

The pull request is created to the release-6.0.0-dev. 
Now it uses Finbuckle.MultiTenant project, later it can be replaced with NuGet package.
The Blazor sample project is added to samples\ASP.NET Core 3: DataIsolationBlazorSample
It is based on the existing DataIsolationSample project with some new features.
The example uses multitenant strategies: host and route, which can be managed in appsettings.json.
The section:
```
"TenantStrategy": {
    "RouteStrategy": false,
    "DefaultRoute": "{__tenant__=}/{controller=Home}/{action=Index}",
    "HostTemplate": "__tenant__.*"
  },
```
When "RouteStrategy": true, the app has three menus
**Megacorp Tenant
Finbuckle Tenant
Initech Tenant**
as in original sample. 
When "RouteStrategy": false the app has one menu 

**Current Tenant**

and shows the db data for the current tenant.

To test host strategy use one of ways:
1. Edit Windows hosts file
	127.0.0.1 megacorp
	127.0.0.1 finbuckle
	127.0.0.1 initech

2. I prefer to use DNS like xip.io
	https://megacorp.127.0.0.1.xip.io:5001
	https://finbuckle.127.0.0.1.xip.io:5001
	https://initech.127.0.0.1.xip.io:5001

**Comment by natelaff** [comment]
One thing I'm running into here is with blazor server, basepath strategy seems to have issues. On some calls I end up with '_blazor' as the attempted tenant name which causes things to bomb. Has anyone figured this out?

**Comment by AndrewTriesToCode** [comment]
@natelaff is `_blazor` actually he base path of the requests or is Blazor modifying the base path? In `AddMultiTenant` you can specifiy identifiers to ignore, but I'm not sure if that will work. Might have to reimplement your own strategy to handle this odd behavior.

**Comment by natelaff** [comment]
yeah it seems to be modifying it on its own. i will try with the ignore identifiers, i missed that bit. honestly, after three days about to ditch blazor efforts. the oidc picture there is not good.

**Comment by riccardone** [comment]
Thanks @Oleg26Dev for the sample code. 
I'm trying to use it with the nuget instead of direct reference to Finbuckle.MultiTenant project. 

To do so I remove the direct reference to Finbuckle.MultiTenant project and instead add the following nugets
Finbuckle.MultiTenant
Finbuckle.MultiTenant.AspNetCore
Finbuckle.MultiTenant.EntityFrameworkCore

but there are a bunch of extensions and related classes that seems to be missed. 
@AndrewTriesToCode is there an additional nuget that I'm missing or is this something still on development?

Many thanks

**Comment by jaltenbernd** [comment]
Trying an implementation of what @Oleg26Dev put together in his sample code into a **.NET 6 Blazor Server application**. Able to get it to work when a Controller is added. But, the HttpContextAccessor.GetMultiTenant() is always null. How does Blazor's HttpContextAccessor communicate to the GetMultiTenant()? 

My setup is using the EFStore with RouteStrategy. 

<pre>
//Allows accessing HttpContext in Blazor
builder.Services.AddHttpContextAccessor();

builder.Services.AddScoped<ContextHelper>();

builder.Services.AddMultiTenant<TenantInfo>()
    .WithEFCoreStore<AppDbContext, TenantInfo>()
    .WithRouteStrategy()
    .WithPerTenantOptions<CookieAuthenticationOptions>((options, tenantInfo) =>
    {
        // Since we are using the route strategy configure each tenant
        // to have a different cookie name and adjust the paths.
        options.Cookie.Path = $"/{tenantInfo.Identifier}";
        options.Cookie.Name = $"{tenantInfo.Id}_authentication";
        options.LoginPath = $"{options.Cookie.Path}{options.LoginPath}";
        options.LogoutPath = $"{options.Cookie.Path}{options.LogoutPath}";
    });

...

app.UseMultiTenant();
app.MapBlazorHub();
app.MapControllerRoute("default", "{__tenant__=}/{controller=Home}/{action=Index}");
app.MapFallbackToPage("/_Host");
</pre>

Then, when I use the same **ContextHelper** class, the **GetMultiTenantContext** is null. 

<pre>private readonly IHttpContextAccessor _accessor;

public ContextHelper(IHttpContextAccessor accessor)
{
     _accessor = accessor;
}

public TenantInfo GetCurrentTenant()
{
      var context = _accessor.HttpContext;
      var ti = context.GetMultiTenantContext<TenantInfo>().TenantInfo;
      return ti;
}
</pre>

Am I missing something in the **Progam.cs** class?

**Comment by VictorioBerra** [comment]
@jaltenbernd I dont have an answer for your main question, Andrew might pop in an answer but I wanted to let you know the source code is pretty straight forward regarding the HttpContext.

https://github.com/Finbuckle/Finbuckle.MultiTenant/blob/main/src/Finbuckle.MultiTenant.AspNetCore/Extensions/HttpContextExtensions.cs#L18

Ultimately, GetIdentifier on the strategy is called https://github.com/Finbuckle/Finbuckle.MultiTenant/blob/main/src/Finbuckle.MultiTenant.AspNetCore/Strategies/RouteStrategy.cs#L24

This might lead you down the right road.

Worst case scenario, you could clone and reference the code and actually debug and step through everything.

**Comment by AndrewTriesToCode** [comment]
Hello, the short answer is Blazor support isn't quite where I want it to be yet. The route strategy does not coexist well with it.

However, the base path strategy can work. Try this for your setup. A few things to note: 
1. `WithBasePathStrategy(options => options.RebaseAspNetCorePathBase = true);` the true option is needed for this to work
2. `UseStaticFiles` must come after `UseMultiTenant`
3. In your component don't inject `IHttpContextAccessor` but inject `IMultiTenantContextAccessor<T>` or your tenant info directly.
4. If you want to have non-tenant content I recommend you create a default tenant and use it with `WithStaticStrategy` after `WithBasePathStrategy`. This will ensure a tenant always is there so you don't have to constantly check it for null and allow for a homepage or similar to work where no tenant is needed.

Hope this helps!

```cs
using BlazorServerTest.Data;
using Finbuckle.MultiTenant;

var builder = WebApplication.CreateBuilder(args);

// Add services to the container.
builder.Services.AddRazorPages();
builder.Services.AddServerSideBlazor();
builder.Services.AddSingleton<WeatherForecastService>();

builder.Services.AddMultiTenant<TenantInfo>()
    .WithConfigurationStore()
    .WithBasePathStrategy(options => options.RebaseAspNetCorePathBase = true);

var app = builder.Build();

// Configure the HTTP request pipeline.
if (!app.Environment.IsDevelopment())
{
    app.UseExceptionHandler("/Error");
    // The default HSTS value is 30 days. You may want to change this for production scenarios, see https://aka.ms/aspnetcore-hsts.
    app.UseHsts();
}

app.UseHttpsRedirection();

app.UseMultiTenant();
app.UseStaticFiles();

app.UseRouting();

app.MapBlazorHub();
app.MapFallbackToPage("/_Host");

app.Run();
```

**Comment by jaltenbernd** [comment]
Did what you recommended above, but my `IMultiTenantContextAccessor<TenantInfo>` that I inject into the page has the `MultiTenantContext` as null.

**Comment by AndrewTriesToCode** [comment]
Hi @jaltenbernd, can you paste your configuration or link to a repo that I can look at?

**Comment by marcmontecalvo** [comment]
Anyone looking for a pretty good example of how to use Finbuckle with Blazor can check these out (they helped me a ton):
https://fullstackhero.net/blazor-webassembly-boilerplate/general/getting-started/
https://fullstackhero.net/dotnet-webapi-boilerplate/general/getting-started/

**Comment by jaltenbernd** [comment]
@marcmontecalvo How? Your two posts are just general information about Blazor. How does it help with Finbuckle?

**Comment by natelaff** [comment]
@AndrewTriesToCode 
> Hello, the short answer is Blazor support isn't quite where I want it to be yet. The route strategy does not coexist well with it.

Is this something that needs updated in this repo or is it a Blazor issue? I haven't dug into Blazor yet, but am about to dive head first and know I need this working haha. Do you need help with this? I can try my best.

**Comment by AndrewTriesToCode** [comment]
@natelaff glad to here from you. It's mainly a priority / time issue I think.

The routing in Blazor as described [here](https://learn.microsoft.com/en-us/dotnet/api/microsoft.aspnetcore.components.navigationmanager.locationchanged?view=aspnetcore-7.0) does have an event to hook into the route changing, but it's not part of the normal request "middleware" that Finbuckle normally uses for ASP.NET Core when setting the tenant for a request. That being said there's no reason that `TenantResolver` and `IMultiTenantContextAccessor<T>` can't be used here to set the tenant--that's what I would try first.
